data/builder/directory.go

```diff
@@ -0,0 +1,123 @@
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/multiformats/go-multihash"
+)
+
+// https://github.com/ipfs/go-ipfs/pull/8114/files#diff-eec963b47a6e1080d9d8023b4e438e6e3591b4154f7379a7e728401d2055374aR319
+const shardSplitThreshold = 262144
+
+// https://github.com/ipfs/go-unixfs/blob/ec6bb5a4c5efdc3a5bce99151b294f663ee9c08d/io/directory.go#L29
+const defaultShardWidth = 256
+
+// BuildUnixFSRecursive returns a link pointing to the UnixFS node representing
+// the file or directory tree pointed to by `root`
+func BuildUnixFSRecursive(root string, ls *ipld.LinkSystem) (ipld.Link, uint64, error) {
+	info, err := os.Lstat(root)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	m := info.Mode()
+	switch {
+	case m.IsDir():
+		entries, err := os.ReadDir(root)
+		if err != nil {
+			return nil, 0, err
+		}
+		lnks := make([]dagpb.PBLink, 0, len(entries))
+		for _, e := range entries {
+			lnk, sz, err := BuildUnixFSRecursive(path.Join(root, e.Name()), ls)
+			if err != nil {
+				return nil, 0, err
+			}
+			entry, err := BuildUnixFSDirectoryEntry(e.Name(), int64(sz), lnk)
+			if err != nil {
+				return nil, 0, err
+			}
+			lnks = append(lnks, entry)
+		}
+		outLnk, err := BuildUnixFSDirectory(lnks, ls)
+		return outLnk, 0, err
+	case m.Type() == fs.ModeSymlink:
+		content, err := os.Readlink(root)
+		if err != nil {
+			return nil, 0, err
+		}
+		return BuildUnixFSSymlink(content, ls)
+	case m.IsRegular():
+		fp, err := os.Open(root)
+		if err != nil {
+			return nil, 0, err
+		}
+		defer fp.Close()
+		return BuildUnixFSFile(fp, "", ls)
+	default:
+		return nil, 0, fmt.Errorf("cannot encode non regular file: %s", root)
+	}
+}
+
+// estimateDirSize estimates if a directory is big enough that it warrents sharding.
+// The estimate is the sum over the len(linkName) + bytelen(linkHash)
+// https://github.com/ipfs/go-unixfs/blob/master/io/directory.go#L152-L162
+func estimateDirSize(entries []dagpb.PBLink) int {
+	s := 0
+	for _, e := range entries {
+		lnk := e.Hash.Link()
+		cl, ok := lnk.(cidlink.Link)
+		if ok {
+			s += len(e.Name.Must().String()) + cl.ByteLen()
+		} else {
+			s += len(e.Name.Must().String()) + len(lnk.String())
+		}
+	}
+	return s
+}
+
+// BuildUnixFSDirectory creates a directory link over a collection of entries.
+func BuildUnixFSDirectory(entries []dagpb.PBLink, ls *ipld.LinkSystem) (ipld.Link, error) {
+	if estimateDirSize(entries) > shardSplitThreshold {
+		return BuildUnixFSShardedDirectory(defaultShardWidth, multihash.MURMUR3_128, entries, ls)
+	}
+	ufd, err := BuildUnixFS(func(b *Builder) {
+		DataType(b, data.Data_Directory)
+	})
+	if err != nil {
+		return nil, err
+	}
+	pbb := dagpb.Type.PBNode.NewBuilder()
+	pbm, err := pbb.BeginMap(2)
+	if err != nil {
+		return nil, err
+	}
+	pbm.AssembleKey().AssignString("Data")
+	pbm.AssembleValue().AssignBytes(data.EncodeUnixFSData(ufd))
+	pbm.AssembleKey().AssignString("Links")
```

Comment:
I'm a bit uneasy about ignoring errors on these IPLD node calls. If we think they should never error, then I'd `panic` when they do error. Silently ignoring the errors could lead to unforeseen problems in the future, e.g. imagine if IPLD codegen or bindnode are buggy.

Reply:
Entirely reasonable if you want to avoid duplication with a form of `must` helper func, too